flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala

```diff
@@ -800,6 +794,66 @@ class Table(
   }
 
   /**
+    * Writes the [[Table]] to a [[TableSink]] specified by given name. The tableSink name
+    * represents a registered [[TableSink]] which defines an external storage location.
+    *
+    * A batch [[Table]] can only be written to a
+    * [[org.apache.flink.table.sinks.BatchTableSink]], a streaming [[Table]] requires a
+    * [[org.apache.flink.table.sinks.AppendStreamTableSink]], a
+    * [[org.apache.flink.table.sinks.RetractStreamTableSink]], or an
+    * [[org.apache.flink.table.sinks.UpsertStreamTableSink]].*
+    *
+    * @param tableName Name of the [[TableSink]] to which the [[Table]] is written.
+    * @tparam T The data type that the [[TableSink]] expects.
+    */
+  def insertInto[T](tableName: String): Unit = {
+    insertInto(tableName, QueryConfig.getQueryConfigFromTableEnv(this.tableEnv))
+  }
+
+  /**
+    * Writes the [[Table]] to a [[TableSink]] specified by given name. The tableSink name
+    * represents a registered [[TableSink]] which defines an external storage location.
+    *
+    * A batch [[Table]] can only be written to a
+    * [[org.apache.flink.table.sinks.BatchTableSink]], a streaming [[Table]] requires a
+    * [[org.apache.flink.table.sinks.AppendStreamTableSink]], a
+    * [[org.apache.flink.table.sinks.RetractStreamTableSink]], or an
+    * [[org.apache.flink.table.sinks.UpsertStreamTableSink]].*
+    *
+    * @param tableName Name of the [[TableSink]] to which the [[Table]] is written.
+    * @param conf The [[QueryConfig]] to use.
+    * @tparam T The data type that the [[TableSink]] expects.
+    */
+  def insertInto[T](tableName: String, conf: QueryConfig): Unit = {
+    require(tableName != null && !tableName.isEmpty, "tableSink must not be null or empty.")
+    // validate if the tableSink is registered
+    if (!tableEnv.isRegistered(tableName)) {
+      throw TableException("tableSink must be registered.")
+    }
+    // find if the tableSink is registered //, include validation internally
+    tableEnv.getTable(tableName) match {
+      case sink: TableSinkTable[_] => {
+        // get row type info of upstream table
+        val rowType = getRelNode.getRowType
+        val srcFieldTypes: Array[TypeInformation[_]] = rowType.getFieldList.asScala
+          .map(field => FlinkTypeFactory.toTypeInfo(field.getType)).toArray
+        // column count validation
+        if (srcFieldTypes.length != sink.fieldTypes.length) {
+          throw TableException(s"source column count doesn't match target table[$tableName]'s.")
+        }
+        // column type validation, no need to validate field names
+        if (sink.fieldTypes.zipWithIndex.exists(f => f._1 != srcFieldTypes(f._2))) {
+          throw TableException(s"source row type doesn't match target table[$tableName]'s.")
```

Comment:
Schema of inserted table must exactly match the schema of the target table. Inserted table: [a: INTEGER, b: VARCHAR], Destination table: [a: INTEGER, b: VARCHAR, c: LONG].